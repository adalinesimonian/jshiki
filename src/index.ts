import Evaluator, {
  UnaryOperatorOptions,
  BinaryOperatorOptions,
  LogicalOperatorOptions,
  OperatorOptions,
  EvaluatorOptions,
} from './evaluator'
import getRuleProxy from './rule-proxy'
import RuleTree, {
  AccessPath,
  AccessRule,
  AllowAccessRule,
  BlockAccessRule,
} from './rule-tree'

export {
  AccessPath,
  AccessRule,
  AllowAccessRule,
  BlockAccessRule,
  UnaryOperatorOptions,
  BinaryOperatorOptions,
  LogicalOperatorOptions,
  OperatorOptions,
}

/**
 * Options for parsing.
 */
export interface JshikiParseOptions extends EvaluatorOptions {
  /**
   * Access rules to use when determining what properties can be accessed by the
   * expression. The rules are evaluated in order.
   * @example
   * ```js
   * const options = {
   *   rules: [
   *     { allow: 'foo.bar' },
   *     { block: ['baz', 'qux'] },
   *   ]
   * }
   * const obj = {
   *   foo: { bar: 'baz' },
   *   baz: { qux: 'quux' },
   * }
   * let expr = jshiki.parse('foo.bar', options)
   * expr(obj) // => 'baz'
   *
   * expr = jshiki.parse('baz.qux', options)
   * expr(obj) // => undefined
   * ```
   */
  rules?: AccessRule[]
  /**
   * If true, only properties with a matching allow rule can be accessed by the
   * expression. If false, all properties can be accessed unless they have a
   * block rule. Defaults to false.
   * @default false
   * @example
   * ```js
   * const options = {
   *   explicitAllow: true,
   *   rules: [{ allow: 'foo.bar' }]
   * }
   * const obj = {
   *   foo: { bar: 'baz' },
   *   baz: { qux: 'quux' },
   * }
   * let expr = jshiki.parse('foo.bar', options)
   * expr(obj) // => 'baz'
   *
   * let expr = jshiki.parse('baz', options)
   * expr(obj) // => undefined
   * ```
   */
  explicitAllow?: boolean
}

/**
 * Options for evaluating.
 */
export interface JshikiEvaluateOptions extends JshikiParseOptions {
  /**
   * The scope to use when evaluating the expression. The expression will be
   * limited to accessing the properties of the scope.
   * @example
   * ```js
   * const options = {
   *   scope: {
   *     foo: 'bar',
   *     baz: {
   *       qux: 'quux',
   *     }
   *   }
   * }
   * jshiki.evaluate('foo', options) // => 'bar'
   * jshiki.evaluate('baz.qux', options) // => 'quux'
   * ```
   */
  scope?: Record<any, any>
}

/**
 * A compiled, executable expression.
 *
 * ```js
 * const expression = jshiki.parse('foo.bar')
 * expression({
 *   foo: { bar: 'baz' },
 * }) // => 'baz'
 * ```
 * @param scope The scope to use when evaluating the expression. The expression
 * will be limited to accessing the properties of the scope.
 */
export type JshikiExpression = (scope?: Record<any, any>) => any

/**
 * Parses an expression into an executable function.
 * @param str The expression to parse.
 * @param options The options to use.
 * @returns The parsed expression.
 * @example
 * ```js
 * const expression = jshiki.parse('(a || 1) + 2')
 * let result = expression()
 * // result => 3
 * result = expression({ a: 5 })
 * // result => 7
 * ```
 */
export function parse(
  str: string,
  { rules, explicitAllow, operators, expressions }: JshikiParseOptions = {}
): JshikiExpression {
  const ruleTree = rules ? new RuleTree(rules) : undefined
  const expression = new Evaluator({ operators, expressions }).createExpression(
    str
  )

  return (scope: Record<any, any> = {}) => {
    const proxiedScope = ruleTree
      ? getRuleProxy(scope, ruleTree, Boolean(explicitAllow))
      : scope
    return expression(proxiedScope)
  }
}

/**
 * Evaluates an expression.
 * @param str The expression to evaluate.
 * @param options The options to use.
 * @returns The result of the expression.
 * @example
 * ```js
 * let result = jshiki.evaluate('1 + 2')
 * // result => 3
 * result = jshiki.evaluate('a + 2', { scope: { a: 5 } })
 * // result => 7
 * ```
 */
export function evaluate(str: string, options?: JshikiEvaluateOptions): any {
  return parse(str, options)(options?.scope)
}

export default { parse, evaluate }
