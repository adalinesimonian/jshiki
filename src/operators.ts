export const unary: { [operator: string]: (o: any) => any } = {
  '+': (o: any): number => +o,
  '-': (o: any): number => -o,
  '!': (o: any): boolean => !o,
  '~': (o: any): number => ~o,
  typeof: (o: any): string => typeof o,
}

export const binary: { [operator: string]: (l: any, r: any) => any } = {
  '+': (l: any, r: any): any => l + r,
  '-': (l: any, r: any): number => l - r,
  '*': (l: any, r: any): number => l * r,
  '**': (l: any, r: any): number => l ** r,
  '/': (l: any, r: any): number => l / r,
  '%': (l: any, r: any): number => l % r,
  '<': (l: any, r: any): boolean => l < r,
  '>': (l: any, r: any): boolean => l > r,
  '<=': (l: any, r: any): boolean => l <= r,
  '>=': (l: any, r: any): boolean => l >= r,
  /* eslint-disable eqeqeq */
  '==': (l: any, r: any): boolean => l == r,
  '!=': (l: any, r: any): boolean => l != r,
  /* eslint-enable eqeqeq */
  '===': (l: any, r: any): boolean => l === r,
  '!==': (l: any, r: any): boolean => l !== r,
  '|': (l: any, r: any): number => l | r,
  '^': (l: any, r: any): number => l ^ r,
  '&': (l: any, r: any): number => l & r,
  '<<': (l: any, r: any): number => l << r,
  '>>': (l: any, r: any): number => l >> r,
  '>>>': (l: any, r: any): number => l >>> r,
  in: (l: any, r: any): boolean => l in r,
  instanceof: (l: any, r: any): boolean => l instanceof r,
}

export const logical: {
  [operator: string]: {
    (l: () => any, r: () => any): any
    async: (
      l: () => Promise<{ value: any }>,
      r: () => Promise<{ value: any }>
    ) => Promise<{ result: any }>
  }
} = {
  '||': Object.assign((l: () => any, r: () => any): any => l() || r(), {
    async: async (
      l: () => Promise<{ value: any }>,
      r: () => Promise<{ value: any }>
    ) => ({ result: (await l()).value || (await r()).value }),
  }),
  '&&': Object.assign((l: () => any, r: () => any): any => l() && r(), {
    async: async (
      l: () => Promise<{ value: any }>,
      r: () => Promise<{ value: any }>
    ) => ({ result: (await l()).value && (await r()).value }),
  }),
  '??': Object.assign((l: () => any, r: () => any): any => l() ?? r(), {
    async: async (
      l: () => Promise<{ value: any }>,
      r: () => Promise<{ value: any }>
    ) => ({ result: (await l()).value ?? (await r()).value }),
  }),
}

export default {
  unary,
  binary,
  logical,
}
