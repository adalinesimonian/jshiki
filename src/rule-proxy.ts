import RuleTree, { AccessPathArray, AccessPathPart } from './rule-tree'

function hasAccess(
  ruleTree: RuleTree,
  explicitAllow: boolean,
  propertyScope: AccessPathArray
) {
  const match = ruleTree.match(propertyScope)
  return explicitAllow
    ? match === true || (typeof match === 'string' && match === 'allow')
    : !(typeof match === 'string' && match === 'block')
}

function getProxyHandler<T extends object>(
  ruleTree: RuleTree,
  explicitAllow: boolean,
  scope: AccessPathArray = []
): ProxyHandler<T> {
  const cache = new Map<AccessPathPart, any>()

  return {
    get(target: T, property: AccessPathPart) {
      if (cache.has(property)) {
        return cache.get(property)
      }

      const propertyScope = scope.concat(property)

      if (!hasAccess(ruleTree, explicitAllow, propertyScope)) {
        cache.set(property, undefined)
        return undefined
      }

      const propValue = (target as any)[property]
      const returnValue =
        (propValue && typeof propValue === 'object') ||
        typeof propValue === 'function'
          ? new Proxy(
              (target as any)[property],
              getProxyHandler(ruleTree, explicitAllow, propertyScope)
            )
          : propValue
      cache.set(property, returnValue)
      return returnValue
    },

    set(_target: T, _property: AccessPathPart, value: any) {
      return value
    },
  }
}

export default function getRuleProxy<T extends object>(
  target: T,
  ruleTree: RuleTree,
  explicitAllow?: boolean
): T {
  return (target && typeof target === 'object') || typeof target === 'function'
    ? new Proxy(target, getProxyHandler(ruleTree, Boolean(explicitAllow)))
    : target
}
