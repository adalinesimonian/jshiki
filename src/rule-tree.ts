export type AccessPathPart = string | symbol
export type AccessPathArray = (string | symbol)[]
/**
 * A path to a property. Can be a string of format `foo.bar.baz` or an array of
 * format `['foo', 'bar', Symbol('baz')]`. '*' can be used as a wildcard. '**'
 * can be used as a wildcard for properties at any depth.
 */
export type AccessPath = string | AccessPathArray
export type AllowAccessRule = {
  /** The path to the property to which access should be allowed. */
  allow: AccessPath
}
export type BlockAccessRule = {
  /** The path to the property to which access should be blocked. */
  block: AccessPath
}
export type AccessRule = AllowAccessRule | BlockAccessRule
type Action = 'allow' | 'block'
type MatchResult = Action | boolean

function normalizePath(path: AccessPath): AccessPathArray {
  if (typeof path === 'string') {
    return path.split('.')
  }
  if (!Array.isArray(path)) {
    throw new Error(
      `Path must be a string or an array of strings and/or symbols, got ${typeof path}`
    )
  }
  return path.map(part => {
    if (typeof part !== 'symbol' && typeof part !== 'string') {
      throw new Error(
        `Path part must be a string or a symbol, got ${typeof part}`
      )
    }
    return typeof part === 'symbol' ? part : `${part}`
  })
}

type RuleTreeNodeBase = {
  children: Map<AccessPathPart, RuleTreeNode>
  hasAllowDescendant?: boolean
}

type RuleTreeActionNode = RuleTreeNodeBase & {
  children: Map<AccessPathPart, RuleTreeNode>
  action: Action
  precedence: number
}

type RuleTreeNode = RuleTreeNodeBase | RuleTreeActionNode

function isAllowRule(rule: AccessRule): rule is AllowAccessRule {
  return Object.prototype.hasOwnProperty.call(rule, 'allow')
}

function isActionNode(rule: RuleTreeNode): rule is RuleTreeActionNode {
  return Object.prototype.hasOwnProperty.call(rule, 'action')
}

function unescapeWildcard(part: AccessPathPart): AccessPathPart {
  if (typeof part === 'symbol') {
    return part
  }
  if (part === '\\*') {
    return '*'
  }
  if (part === '\\**') {
    return '**'
  }
  return part
}

export default class RuleTree {
  #rootNodes: Map<AccessPathPart, RuleTreeNode> = new Map()
  #nextPrecedence = 0

  constructor(rules?: AccessRule[]) {
    if (rules) {
      for (const rule of rules) {
        this.add(rule)
      }
    }
  }

  add(rule: AccessRule): RuleTree {
    if (!rule) {
      return this
    }
    const [action, normalized]: [Action, AccessPathArray] = isAllowRule(rule)
      ? ['allow', normalizePath(rule.allow)]
      : ['block', normalizePath(rule.block)]
    if (normalized.length === 0) {
      return this
    }
    let nodeMap = this.#rootNodes
    const precedence = this.#nextPrecedence++
    for (const [index, path] of normalized.entries()) {
      const isLastPart = index === normalized.length - 1
      const existing = nodeMap.get(path)
      if (!existing) {
        if (isLastPart) {
          nodeMap.set(path, {
            children: new Map(),
            action,
            precedence,
          })
          break
        }
        const newNodeMap = new Map()
        nodeMap.set(path, {
          children: newNodeMap,
        })
        nodeMap = newNodeMap
        continue
      }
      if (isLastPart) {
        const existingActionNode = existing as RuleTreeActionNode
        existingActionNode.action = action
        existingActionNode.precedence = precedence
        if (action === 'block') {
          existing.children.clear()
        }
        break
      }
      nodeMap = existing.children
    }
    for (const rootNode of this.#rootNodes.values()) {
      this.#updateHasAllowDescendant(rootNode)
    }
    return this
  }

  #updateHasAllowDescendant(node: RuleTreeNode): void {
    let hasAllowDescendant = false
    for (const [, child] of node.children) {
      this.#updateHasAllowDescendant(child)
      if (
        child.hasAllowDescendant ||
        (child as RuleTreeActionNode).action === 'allow'
      ) {
        hasAllowDescendant = true
      }
    }
    node.hasAllowDescendant = hasAllowDescendant
  }

  /**
   * Checks if the given path has a matching rule. If it does, returns the
   * matching action. If it doesn't, returns a boolean indicating whether or not
   * there is a rule that would allow access to a descendant of the given path.
   * @param path The path to check.
   * @returns If a match is found, the action, either `'allow'` or `'block'`. If
   * no match is found, a boolean indicating whether or not there is a rule that
   * would allow access to a descendant of the given path.
   * @example
   * ```js
   * const ruleTree = new RuleTree([
   *   { allow: 'a.b.c' },
   *   { block: 'x.y.z' },
   * ])
   * let match = ruleTree.match('a.b.c')
   * // match => 'allow'
   * match = ruleTree.match('a.b.d')
   * // match => false
   * match = ruleTree.match('a.b')
   * // match => true
   * match = ruleTree.match('x.y.z')
   * // match => 'block'
   * match = ruleTree.match('x.y.z.alpha')
   * // match => false
   * ```
   */
  match(path: AccessPath): MatchResult {
    const normalized = normalizePath(path)
    if (normalized.length === 0) {
      return false
    }

    let currentNodes = [...this.#rootNodes]
    let nextNodes: [AccessPathPart, RuleTreeNode][] = []
    const actions: Map<
      RuleTreeNode,
      [precedence: number, index: number, action: Action]
    > = new Map()
    let hasAllowDescendant = false

    for (const [index, pathPart] of normalized.entries()) {
      hasAllowDescendant = false

      for (const [nodePathPart, node] of currentNodes) {
        if (nodePathPart === '**') {
          const matchingChild = node.children.get(pathPart)
          const matchingNode = matchingChild ?? node

          if (isActionNode(matchingNode)) {
            actions.set(matchingNode, [
              matchingNode.precedence,
              index,
              matchingNode.action,
            ])
          }

          if (matchingChild) {
            if (matchingChild.hasAllowDescendant) {
              hasAllowDescendant = true
            }

            for (const child of matchingChild.children) {
              nextNodes.push(child)
            }
          } else {
            nextNodes.push([nodePathPart, node])

            if (node.hasAllowDescendant) {
              hasAllowDescendant = true
            }
          }
        }

        if (
          nodePathPart === '*' ||
          unescapeWildcard(nodePathPart) === pathPart
        ) {
          if (isActionNode(node)) {
            actions.set(node, [node.precedence, index, node.action])
          }
          if (node.hasAllowDescendant) {
            hasAllowDescendant = true
          }
          for (const child of node.children) {
            nextNodes.push(child)
          }
        }
      }

      if (nextNodes.length === 0) {
        break
      }

      currentNodes = nextNodes
      nextNodes = []
    }

    if (actions.size > 0) {
      if (actions.size === 1) {
        const action = actions.values().next().value
        return action[1] === normalized.length - 1
          ? action[2]
          : hasAllowDescendant
      }
      // Pick the action with the highest precedence
      let highestPrecedence = -1
      let highestPrecedenceResult: [number, Action]

      const actionArray = [...actions.values()]

      while (actionArray.length > 0) {
        const [precedence, index, action] = actionArray.pop()!
        if (precedence > highestPrecedence) {
          highestPrecedence = precedence
          highestPrecedenceResult = [index, action]
        }
      }

      const [index, action] = highestPrecedenceResult!

      return index === normalized.length - 1 ? action : hasAllowDescendant
    }

    return hasAllowDescendant
  }
}
