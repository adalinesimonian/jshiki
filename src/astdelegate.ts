import {
  ArrayExpression,
  BinaryExpression,
  CallExpression,
  ConditionalExpression,
  Delegate,
  Expression,
  IdentifierExpression,
  LiteralExpression,
  MemberExpression,
  ObjectExpression,
  PropertyExpression,
  Syntax,
  Token,
  TopLevelExpression,
  UnaryExpression,
} from './lib/esprima/types'
import operators from './operators'

export interface ASTDelegateOptions {
  /**
   * The scope to use when evaluating the expression. The expression will be
   * limited to accessing the properties of the scope.
   */
  scope?: Record<any, any>
}

export default class ASTDelegate implements Delegate {
  scope: any

  constructor({ scope }: ASTDelegateOptions = {}) {
    this.scope = scope || {}
  }

  #evaluateIdentifier(expression: IdentifierExpression): any {
    const { identifier } = expression
    if (typeof identifier === 'string') {
      return this.scope[identifier]
    }
    return identifier
  }

  createUnaryExpression(op: string, arg: Expression): UnaryExpression {
    if (!operators.unary[op as keyof typeof operators.unary]) {
      throw Error(`Disallowed operator: ${op}`)
    }

    return Object.assign(
      () => operators.unary[op as keyof typeof operators.unary](arg()),
      {
        type: Syntax.UnaryExpression as const,
        operator: op,
        argument: arg,
      }
    )
  }

  createBinaryExpression(
    op: string,
    left: Expression,
    right: Expression
  ): BinaryExpression {
    if (!operators.binary[op as keyof typeof operators.binary]) {
      throw Error(`Disallowed operator: ${op}`)
    }

    const leftExpr =
      left.type === Syntax.Identifier
        ? () => this.#evaluateIdentifier(left)
        : left
    const rightExpr =
      right.type === Syntax.Identifier
        ? () => this.#evaluateIdentifier(right)
        : right

    return Object.assign(
      () =>
        operators.binary[op as keyof typeof operators.binary](
          leftExpr(),
          rightExpr()
        ),
      {
        type: Syntax.BinaryExpression as const,
        operator: op,
        left,
        right,
      }
    )
  }

  createConditionalExpression(
    test: Expression,
    consequent: Expression,
    alternate: Expression
  ): ConditionalExpression {
    return Object.assign(() => (test() ? consequent() : alternate()), {
      type: Syntax.ConditionalExpression as const,
      test,
      consequent,
      alternate,
    })
  }

  createIdentifier(identifier: string | boolean | null): IdentifierExpression {
    return Object.assign(() => identifier, {
      type: Syntax.Identifier as const,
      identifier: identifier,
    })
  }

  createMemberExpression(
    accessor: '.' | '[',
    object: Expression,
    property: IdentifierExpression
  ): MemberExpression {
    const objectEvaluator =
      object.type === Syntax.Identifier
        ? () => this.#evaluateIdentifier(object)
        : object

    return Object.assign(() => objectEvaluator()[property()], {
      type: Syntax.MemberExpression as const,
      accessor,
      object,
      property,
    })
  }

  createCallExpression(callee: Expression, args: Expression[]): CallExpression {
    return Object.assign(
      () => {
        let func, self
        if (callee.type === Syntax.Identifier) {
          self = this.#evaluateIdentifier(callee)
          func = self
        } else if (callee.type === Syntax.MemberExpression) {
          self = callee.object()
          func = callee()
        } else {
          self = callee()
          func = callee
        }
        return func.apply(
          self,
          args.map(arg => arg())
        )
      },
      {
        type: Syntax.CallExpression as const,
        callee,
        args,
      }
    )
  }

  createLiteral(
    token: Pick<
      Extract<Token, { value: string | number | boolean | null }>,
      'value'
    >
  ): LiteralExpression {
    return Object.assign(() => token.value, {
      type: Syntax.Literal as const,
      value: token.value,
    })
  }

  createArrayExpression(elements: (Expression | null)[]): ArrayExpression {
    return Object.assign(() => elements.map(element => element && element()), {
      type: Syntax.ArrayExpression as const,
      elements,
    })
  }

  createProperty(
    kind: 'init',
    key: IdentifierExpression,
    value: Expression
  ): PropertyExpression {
    return Object.assign(
      () => ({
        key: key({ child: true }),
        value: value(),
      }),
      {
        type: Syntax.Property as const,
        kind,
        key,
        value,
      }
    )
  }

  createObjectExpression(properties: Expression[]): ObjectExpression {
    return Object.assign(
      () => {
        const object: Record<any, any> = {}
        for (const property of properties) {
          const { key, value } = property()
          object[key] = value
        }
        return object
      },
      {
        type: Syntax.ObjectExpression as const,
        properties,
      }
    )
  }

  createTopLevel(expression: Expression): TopLevelExpression {
    let evaluator
    if (expression.type === Syntax.Identifier) {
      const identifier = this.#evaluateIdentifier(expression)
      evaluator = () => identifier
    } else {
      evaluator = expression
    }

    return Object.assign(evaluator, {
      type: Syntax.TopLevel as const,
      expression,
    })
  }

  createThisExpression(): never {
    // TODO
    throw new Error('`this` is not supported')
  }
}
