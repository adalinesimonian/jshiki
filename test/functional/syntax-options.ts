import { evaluate, evaluateAsync, parse, parseAsync } from '../../src'

describe('Syntax options', () => {
  it('should allow all supported syntax by default', () => {
    expect(evaluate('a.b.c', { scope: { a: { b: { c: 1 } } } })).toBe(1)
    expect(evaluate('a?.b?.c', { scope: { a: { b: { c: 1 } } } })).toBe(1)
    expect(evaluate('a["b"]["c"]', { scope: { a: { b: { c: 1 } } } })).toBe(1)
    expect(evaluate('a?.["b"]?.["c"]', { scope: { a: { b: { c: 1 } } } })).toBe(
      1
    )
    expect(evaluate('a()', { scope: { a: () => 1 } })).toBe(1)
    expect(evaluate('a?.()', { scope: { a: () => 1 } })).toBe(1)
    expect(
      // eslint-disable-next-line no-template-curly-in-string
      evaluate('tag`a${b}`', {
        scope: {
          b: 1,
          tag: (strings: TemplateStringsArray, expression: number) =>
            `got ${strings.join('')} and ${expression}`,
        },
      })
    ).toBe('got a and 1')
    // eslint-disable-next-line no-template-curly-in-string
    expect(evaluate('`a ${b}`', { scope: { b: 1 } })).toBe('a 1')
    expect(evaluate('{ a: 1 }')).toEqual({ a: 1 })
    // eslint-disable-next-line no-sparse-arrays
    expect(evaluate('[1, , 3]')).toEqual([1, , 3])
    expect(evaluate('/x/g')).toEqual(/x/g)
  })

  it('should allow syntax when corresponding option is set to true', () => {
    const options = {
      syntax: {
        memberAccess: true,
        calls: true,
        taggedTemplates: true,
        templates: true,
        objects: true,
        arrays: true,
        regexes: true,
      },
    }

    expect(
      evaluate('a.b.c', { ...options, scope: { a: { b: { c: 1 } } } })
    ).toBe(1)
    expect(
      evaluate('a?.b?.c', { ...options, scope: { a: { b: { c: 1 } } } })
    ).toBe(1)
    expect(
      evaluate('a["b"]["c"]', { ...options, scope: { a: { b: { c: 1 } } } })
    ).toBe(1)
    expect(
      evaluate('a?.["b"]?.["c"]', {
        ...options,
        scope: { a: { b: { c: 1 } } },
      })
    ).toBe(1)
    expect(evaluate('a()', { ...options, scope: { a: () => 1 } })).toBe(1)
    expect(evaluate('a?.()', { ...options, scope: { a: () => 1 } })).toBe(1)
    expect(
      // eslint-disable-next-line no-template-curly-in-string
      evaluate('tag`a${b}`', {
        ...options,
        scope: {
          b: 1,
          tag: (strings: TemplateStringsArray, expression: number) =>
            `got ${strings.join('')} and ${expression}`,
        },
      })
    ).toBe('got a and 1')
    // eslint-disable-next-line no-template-curly-in-string
    expect(evaluate('`a ${b}`', { ...options, scope: { b: 1 } })).toBe('a 1')
    expect(evaluate('{ a: 1 }', options)).toEqual({ a: 1 })
    // eslint-disable-next-line no-sparse-arrays
    expect(evaluate('[1, , 3]', options)).toEqual([1, , 3])
    expect(evaluate('/x/g')).toEqual(/x/g)
  })

  it('should block syntax when corresponding option is set to false', () => {
    const options = {
      syntax: {
        memberAccess: false,
        calls: false,
        taggedTemplates: false,
        templates: false,
        objects: false,
        arrays: false,
        regexes: false,
      },
    }

    expect(() => parse('a.b.c', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('a?.b?.c', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('a["b"]["c"]', options)).toThrowErrorMatchingSnapshot()
    expect(() =>
      parse('a?.["b"]?.["c"]', options)
    ).toThrowErrorMatchingSnapshot()
    expect(() => parse('a()', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('a?.()', options)).toThrowErrorMatchingSnapshot()
    expect(() =>
      // eslint-disable-next-line no-template-curly-in-string
      parse('tag`a${b}`', options)
    ).toThrowErrorMatchingSnapshot()
    expect(() =>
      // eslint-disable-next-line no-template-curly-in-string
      parse('`a ${b}`', options)
    ).toThrowErrorMatchingSnapshot()
    expect(() => parse('{ a: 1 }', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('[1, , 3]', options)).toThrowErrorMatchingSnapshot()
    expect(() => parse('/x/g', options)).toThrowErrorMatchingSnapshot()
  })
})

describe('Syntax options (async)', () => {
  it('should allow all supported syntax by default', async () => {
    expect(
      await evaluateAsync('a.b.c', { scope: { a: { b: { c: 1 } } } })
    ).toBe(1)
    expect(
      await evaluateAsync('a?.b?.c', { scope: { a: { b: { c: 1 } } } })
    ).toBe(1)
    expect(
      await evaluateAsync('a["b"]["c"]', { scope: { a: { b: { c: 1 } } } })
    ).toBe(1)
    expect(
      await evaluateAsync('a?.["b"]?.["c"]', { scope: { a: { b: { c: 1 } } } })
    ).toBe(1)
    expect(await evaluateAsync('a()', { scope: { a: () => 1 } })).toBe(1)
    expect(await evaluateAsync('a?.()', { scope: { a: () => 1 } })).toBe(1)
    expect(
      // eslint-disable-next-line no-template-curly-in-string
      await evaluateAsync('tag`a${b}`', {
        scope: {
          b: 1,
          tag: (strings: TemplateStringsArray, expression: number) =>
            `got ${strings.join('')} and ${expression}`,
        },
      })
    ).toBe('got a and 1')
    // eslint-disable-next-line no-template-curly-in-string
    expect(await evaluateAsync('`a ${b}`', { scope: { b: 1 } })).toBe('a 1')
    expect(await evaluateAsync('{ a: 1 }')).toEqual({ a: 1 })
    // eslint-disable-next-line no-sparse-arrays
    expect(await evaluateAsync('[1, , 3]')).toEqual([1, , 3])
    expect(await evaluateAsync('/x/g')).toEqual(/x/g)
  })

  it('should allow syntax when corresponding option is set to true', async () => {
    const options = {
      syntax: {
        memberAccess: true,
        calls: true,
        taggedTemplates: true,
        templates: true,
        objects: true,
        arrays: true,
        regexes: true,
      },
    }

    expect(
      await evaluateAsync('a.b.c', {
        ...options,
        scope: { a: { b: { c: 1 } } },
      })
    ).toBe(1)
    expect(
      await evaluateAsync('a?.b?.c', {
        ...options,
        scope: { a: { b: { c: 1 } } },
      })
    ).toBe(1)
    expect(
      await evaluateAsync('a["b"]["c"]', {
        ...options,
        scope: { a: { b: { c: 1 } } },
      })
    ).toBe(1)
    expect(
      await evaluateAsync('a?.["b"]?.["c"]', {
        ...options,
        scope: { a: { b: { c: 1 } } },
      })
    ).toBe(1)
    expect(
      await evaluateAsync('a()', { ...options, scope: { a: () => 1 } })
    ).toBe(1)
    expect(
      await evaluateAsync('a?.()', { ...options, scope: { a: () => 1 } })
    ).toBe(1)
    expect(
      // eslint-disable-next-line no-template-curly-in-string
      await evaluateAsync('tag`a${b}`', {
        ...options,
        scope: {
          b: 1,
          tag: (strings: TemplateStringsArray, expression: number) =>
            `got ${strings.join('')} and ${expression}`,
        },
      })
    ).toBe('got a and 1')
    expect(
      // eslint-disable-next-line no-template-curly-in-string
      await evaluateAsync('`a ${b}`', { ...options, scope: { b: 1 } })
    ).toBe('a 1')
    expect(await evaluateAsync('{ a: 1 }', options)).toEqual({ a: 1 })
    // eslint-disable-next-line no-sparse-arrays
    expect(await evaluateAsync('[1, , 3]', options)).toEqual([1, , 3])
    expect(await evaluateAsync('/x/g')).toEqual(/x/g)
  })

  it('should block syntax when corresponding option is set to false', async () => {
    const options = {
      syntax: {
        memberAccess: false,
        calls: false,
        taggedTemplates: false,
        templates: false,
        objects: false,
        arrays: false,
        regexes: false,
      },
    }

    expect(async () =>
      parseAsync('a.b.c', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('a?.b?.c', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('a["b"]["c"]', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('a?.["b"]?.["c"]', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('a()', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('a?.()', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      // eslint-disable-next-line no-template-curly-in-string
      parseAsync('tag`a${b}`', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      // eslint-disable-next-line no-template-curly-in-string
      parseAsync('`a ${b}`', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('{ a: 1 }', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('[1, , 3]', options)
    ).rejects.toThrowErrorMatchingSnapshot()
    expect(async () =>
      parseAsync('/x/g', options)
    ).rejects.toThrowErrorMatchingSnapshot()
  })
})
