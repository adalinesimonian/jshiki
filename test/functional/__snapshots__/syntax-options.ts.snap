// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Syntax options (async) should block syntax when corresponding option is set to false 1`] = `"Member access is not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 2`] = `"Member access is not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 3`] = `"Member access is not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 4`] = `"Member access is not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 5`] = `"Function calls are not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 6`] = `"Function calls are not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 7`] = `"Tagged template literals are not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 8`] = `"Template literals are not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 9`] = `"Object literals are not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 10`] = `"Array literals are not allowed"`;

exports[`Syntax options (async) should block syntax when corresponding option is set to false 11`] = `"Regular expressions are not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 1`] = `"Member access is not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 2`] = `"Member access is not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 3`] = `"Member access is not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 4`] = `"Member access is not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 5`] = `"Function calls are not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 6`] = `"Function calls are not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 7`] = `"Tagged template literals are not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 8`] = `"Template literals are not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 9`] = `"Object literals are not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 10`] = `"Array literals are not allowed"`;

exports[`Syntax options should block syntax when corresponding option is set to false 11`] = `"Regular expressions are not allowed"`;
